### PR TITLE
Bring the loan and hold reapers into the generic reaper system

### PR DESCRIPTION
**Pull request by leonardr**
This branch removes `LoanAndHoldReaper`, replacing it with two `ReaperMonitor`s that fit into the general reaper system. The `loan_reaper` script is replaced with the `database_reaper` script.

The reapers do not affect collections where the circulation manager is the source of truth w/r/t loans and holds. Right now that means collections based on ODL or OPDS For Distributors.

### CODE REVIEW

File: api/opds_for_distributors.py [hunk 1]
```diff
@@ -143,13 +143,23 @@ def checkin(self, patron, pin, licensepool):
             pass
 
     def checkout(self, patron, pin, licensepool, internal_format):
+        now = datetime.datetime.utcnow()
```
Comment (leonardr) [inline api/opds_for_distributors.py]: I also started using utcnow() here instead of now().

**Comment by aslagle** [comment]
I didn't notice these scripts before and wrote a separate Hold reaper for ODL. But now that I'm looking I think this might be a problem. 365 days doesn't seem all that long to have a popular book on hold, and for ODL you can't just sync it from somewhere else if it's accidentally deleted.

**Comment by leonardr** [comment]
This is because in some cases the Loan/Hold object is the source of truth? Can we explicitly mark the source-of-truth items so they don't get reaped?

**Comment by aslagle** [comment]
Maybe we could filter out things that have an external identifier?

File: api/monitor.py [hunk 2]
```diff
@@ -228,3 +234,49 @@ def get_identifiers(self, url=None):
...
+    def where_clause(self):
+        """Find loans/holds that have either expired, or that were created a
+        long time ago and have no definite end date.
+        """
+        cls = self.MODEL_CLASS
+        end_field = getattr(cls, 'end')
+        start_field = getattr(cls, 'start')
+
+        now = datetime.datetime.utcnow()
+        expired = end_field < now
```
Comment (aslagle) [inline api/monitor.py]: I think for a Hold the end date is an estimate until it's available to check out, so an end date slightly in the past doesn't necessarily mean it's expired.

**Comment by leonardr** [comment]
We could use another subquery to make sure license_pool_id is not in the list of license pools whose collection_id is one of the 'source of truth' collections. That might have bad performance characteristics though. The subqueries were the only way I could find to do all the deletes in one SQL statement rather than having to iterate in batches, but it might be faster in the worst case to iterate in batches.

**Comment by aslagle** [comment]
It could be added to the existing subquery. It shouldn't be much worse than filtering on open_access, should it?

**Comment by leonardr** [comment]
It depends on the collection. In a library where loans are being given out for open-access books (as opposed to people just downloading the books), it's likely that open-access books are mixed with non-open access, and there will be relatively few open-access loans. In a library that hosts a source-of-truth collection it seems likely that a significant number of the loans will be for that collection. But since the subquery is a query against Loan, not against LicensePool, it should be manageable.

**Pull request merged**